### PR TITLE
Map legend

### PR DESCRIPTION
**Pull request by Icicleta**
New style of map legend, it just show up in organisation page

### CODE REVIEW

**Comment by tansaku** [comment]
Great work on this @Icicleta !  Very impressed with the good use of CSS and that you got started on exploring how to restrict the map legend to particular pages.

Here are some notes and/or things for you to work before we sync up next.  Please don't worry about how far or how much you do.  Simply trying things out (as you are doing) is exactly the way to go.  I'll give you a hand with cucumber tomorrow if you've gotten stuck with the feature tests :-)

1) So we got a little ahead of ourselves (my fault entirely) by working on some functionality without having worked on a feature test.  I know you're not so comfortable with cucumber, but we have an existing feature test that checks that the map legend appears on the controller page:

features/home_page/map.feature

```
  Scenario: Showing meaning of large map icons
    Given I visit the home page
    And I click "Close"
    Then I should see "Details updated by the organisation within the last 12 months"
    Then I should see "Details NOT updated by the organisation within the last 12 months"
```

I think we could do with a pair of scenarios:

a) Scenario: Showing meaning of large and small map icons on home page
b) Scenario: Do not show meaning of large and small map icons on volunteer pages

scenario a) could just be the one above, but what about scenario b)? :-)

Ideally we might revert the code that we just checked in to make sure that the test will fail initially so that we can then make it pass, but see below for a suggestion for perhaps a cleaner way to make it pass :-)

2) In terms of avoiding replication of templates, I think a better way to go is to keep a single two_columns layout (possibly renaming it two_columns_with_map) and work on splitting out the content in app/views/shared/_map_scripts.html.erb

If you take a look at some of the views for organisations and volunteer_ops you'll see that all of _form.html.erb, index.html.erb and show.html.erb have the following at the end:

```
<% render partial: 'shared/map_scripts', locals: {markers: @markers} %>
```

that pulls in the file  'app/views/shared/_map_scripts.html.erb' which then populates the content markers in the two_column layout.

I have a hypothesis that we could take the map legend code/html from _map_scripts.html.erb and stick it in 'app/views/shared/_map_legend_script.html.erb' and then only pull that in in the organisation views where we want the legend to show up.  Makes sense?

If not I'll follow up with you on slack or in this pull request :-)

File: features/home_page/map.feature [hunk 1]
```diff
@@ -63,4 +63,10 @@ Feature: Map of local charities
     Given I visit the home page
     And I click "Close"
     Then I should see "Details updated by the organisation within the last 12 months"
-    Then I should see "Details NOT updated by the organisation within the last 12 months"
\ No newline at end of file
+    Then I should see "Details NOT updated by the organisation within the last 12 months"
+
+  Scenario: Do not showing meaning of large map icons
```
Comment (tansaku) [inline features/home_page/map.feature]: I would change this scenario name to: "Do not show meaning of large map icons on volunteer ops page"

**Comment by tansaku** [comment]
Hey @icicleta - great work on this!

I've mutation tested the cucumber feature tests and all looks good.

I'd just update the scenario names as follows:

```cucumber
Scenario: Show meaning of large map icons on home page
Scenario: Do not show meaning of large map icons on volunteer ops page
```

The only other thing we might do is discuss the merits of using flexbox vs the bootstrap stuff I had before.  I wonder what the browser compatibility is like for those flexbox tags ...

**Comment by Icicleta** [comment]
Sam it seems right to change scenarios names specifying the pages were is visible and were is not. 

About flex, I thought it was a bootstrap feature, sorry. I used though because it was being used also on the div above 'map_container' in 'two_columns_with_map'. And I thought it was the best to align icon and text. I didn't know though whether key-row was relevant in backend, so I decided to leave it there even I didn't need it in css.
But whatever you think is better please go ahead. Thanks for your comments. :)

**Comment by tansaku** [comment]
thanks for the update @Icicleta I think flex is not part of bootstrap, although someone might correct me :-)

According to http://caniuse.com/#search=flex I think flex is not so well supported in IE and one of the things we have to consider for this site, is that many of our stakeholders may be using older computers and thus older browsers so we are concerned with accessibility issues.

I very much like the style you've created - and I'll show it to the client today, but I think we might want to go with the same using simply bootstrap, or possibly even a table, which I know we all don't like much, although as I recently re-read what tables are supposed to be used for; I start to think that this might actually be a case where a table is justified.

The key thing with tables is, of course, that we should NOT use them for layout, but only for tabular data.  But if we consider what a map legend IS, it's a set of tabular data:

* http://webdesignfromscratch.com/html-css/html-tables/
* http://www.w3.org/TR/html5/tabular-data.html
* https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table

Now I'm not a CSS Guru like @sjmog, so everyone should feel free to correct me, but I'm wondering if there isn't a strong argument for using a table here ...

**Comment by sjmog** [comment]
Hello! It's less important that the semantic HTML used indicates the _kind_ of data than how it fits into the information flow of the page.

So things like `<aside>` are good for non-critical, but helpful, information. So `aside` is an option for a legend.

On the other hand, `<table>` is good for displaying two-dimensional information. If you wrap a `<table>` inside an `<aside>`, I think you get the best of both worlds:

```html
<aside>
  <table>
    ...
  </table>
</aside>
```

As for Flexbox:`flex` is poorly-supported in older version of Internet Explorer so it won't be ready for most people until a year's time (once more people have upgraded).

Until then, if you need to vertically center things, you will need to use something like the techniques discussed [on Chris Coyier's blog, CSS Tricks](https://css-tricks.com/centering-in-the-unknown/).

Does that answer the question?

**Comment by Icicleta** [comment]
Hi Sam's,

Thanks for all this update about flex. @tansaku what I meant with flex and bootstrap wasn't about flex itself, I meant the 'flex-row' class. I explained myself  a bit bad :)

Fair enough that we want to make it work with old versions of IE. So I'll follow @sjmog suggestions, that actually I find very interesting, thanks @sjmog. 

Also I'll have a look to your links @tansaku, find useful knowing when use tables, even when I don' like them ;)

Thanks both of you

**Comment by Icicleta** [comment]
@tansaku, I've just pushed again, so you should have a new pull request. 
Two things,
1. There is a key_title class that is styled, I don't see this class in the html and I didn't find it anywhere searching after press double shift. I have comment it out just in case, if you reckon is useless you can delete it.
2. @sjmog I have used <aside> as you suggested. I would like to ask you whether do you thing is needed to be used. I didn't use it on css so I wonder whether it is more about semantic and for sake of clarity than style issue.

Many thanks

**Comment by tansaku** [comment]
thanks @Icicleta - please merge to develop if you haven't already and I'll check on the train home :-)

**Comment by Icicleta** [comment]
I'm not at home @tansaku,  I will do it this this evening. I have pushed to maps develop though,  in case you wanna have a look :)

**Comment by tansaku** [comment]
@Icicleta cool - can you also sync to develop? many thanks in advance

**Comment by Icicleta** [comment]
@tanzaku, now everything is ok, I've pull develop, and merged map_legend to my develop. Finally pushed to origin develop.

**Comment by tansaku** [comment]
ready to go @Icicleta maybe just remove the commented .key_title css and remove the newline from the start of organisations_controller and I will merge on monday

**Comment by Icicleta** [comment]
great!! give me 5 :)

**Comment by Icicleta** [comment]
I've just pushed again with changes you told me. Newline, yo mean the empty first line of the file?

**Comment by tansaku** [comment]
yes I do :-)

**Comment by tansaku** [comment]
I'll merge this in here and then move to develop myself - probably easier than re-opening and it will maintain all the chat about this feature here

**Pull request merged**